src/auth/index.js

```diff
@@ -274,4 +274,65 @@ export default class Auth {
     this.storage.remove(authCommon.IMPERSONATION_USER_KEY);
     this.storage.remove(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY);
   }
+
+  parseRedirectFragment(fragment, ourState) {
+    // After being redirected from oauth, the URL will look like:
+    // https://todo.examples.stitch.mongodb.com/#_stitch_state=...&_stitch_ua=...
+    // This function parses out stitch-specific tokens from the fragment and
+    // builds an object describing the result.
+    const vars = fragment.split('&');
+    const result = { ua: null, found: false, stateValid: false, lastError: null };
+    let shouldBreak = false;
+    for (let i = 0; i < vars.length; ++i) {
+      const pairParts = vars[i].split('=');
+      const pairKey = decodeURIComponent(pairParts[0]);
+      switch (pairKey) {
+      case authCommon.STITCH_ERROR_KEY:
+        result.lastError = decodeURIComponent(pairParts[1]);
+        result.found = true;
+        shouldBreak = true;
+        break;
+      case authCommon.USER_AUTH_KEY:
+        try {
+          result.ua = this.unmarshallUserAuth(decodeURIComponent(pairParts[1]));
+          result.found = true;
+        } catch (e) {
+          result.lastError = e;
+        }
+        continue;
+      case authCommon.STITCH_LINK_KEY:
+        result.found = true;
+        continue;
+      case authCommon.STATE_KEY:
+        result.found = true;
+        let theirState = decodeURIComponent(pairParts[1]);
+        if (ourState && ourState === theirState) {
+          result.stateValid = true;
+        }
+        continue;
+      default: continue;
+      }
+
+      if (shouldBreak) {
+        break;
+      }
+    }
+
+    return result;
+  }
+
+  unmarshallUserAuth(data) {
+    let parts = data.split('$');
+    if (parts.length !== 4) {
```

Comment:
[nit] might want to put this `4` in a const somewhere higher up in the file, although not super important